### PR TITLE
Page editor side panels

### PR DESCRIPTION
**Pull request by kaedroho**
This PR implements a panel that slides out from the right hand side of the page editor:
image

This is where the info, comments, history and preview screens will go (all are currently blank except for comments).

Populating these side panels with content and styling the comments to look nicer in the panel will be done in subsequent PRs

### CODE REVIEW

File: wagtail/admin/templates/wagtailadmin/shared/breadcrumb-next.html [hunk 1]
```diff
@@ -0,0 +1,67 @@
+{% load i18n wagtailadmin_tags %}
+
+{% with breadcrumb_link_classes='w-flex w-justify-start w-items-center w-text-grey-600 w-no-underline w-outline-offset-inside hover:w-underline hover:w-text-primary' breadcrumb_list_classes='w-overflow-hidden w-transition w-duration-300 w-whitespace-nowrap w-flex-shrink-0' icon_classes='w-w-4 w-h-4 w-mr-3' %}
+    {# Breadcrumbs are visible on mobile by default but hidden on desktop #}
+    <button data-toggle-breadcrumbs  {# handled in page-editor.js #}
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/shared/breadcrumb-next.html:5]: this needs `type="button"` as the default is `submit` which could lead to unexpected behaviours. ref - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button

File: wagtail/admin/templates/wagtailadmin/shared/headers/page_edit_header.html [hunk 2]
```diff
@@ -28,28 +27,10 @@ <h1 class="w-sr-only">
...
-                    <div class="w-w-[10px] w-h-[10px] w-flex w-justify-center w-items-center w-p-[3px] w-font-bold w-absolute w-bg-teal-100 w-text-white w-text-[10px] w-border w-border-white w-rounded-full -w-top-[10px] -w-right-[12px]"
-                        data-comments-toggle-count>
-                        {# populated in page-editor.js #}
-                    </div>
-                {% endif %}
-            </button>
-        {% endwith %}
+        {# Side panel Toggles #}
+        {% for panel in side_panels %}
+            {% component panel.toggle %}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/shared/headers/page_edit_header.html:32]: Could we make those toggles reuse the `nav_icon_classes` and `nav_icon_button_classes` defined on line 16? We need to make sure the buttons have consistent heights:

button-heights

File: wagtail/admin/views/pages/edit.py [hunk 3]
```diff
@@ -32,6 +34,81 @@
...
 
+class BaseSidePanel(Component):
+    def __init__(self, page):
+        self.page = page
+
+    class Toggle(Component):
+        def __init__(self, panel):
+            self.panel = panel
+
+        aria_label = gettext_lazy("Toggle")
```
Comment (thibaudcolas) [inline wagtail/admin/views/pages/edit.py:45]: As far as I can see this default value isn’t used anywhere, so we should remove it to have one less thing to translate. "Toggle" wouldn’t work as an `aria-label`, it’d need to indicate what is being toggled (which we’re doing well in overrides).

File: wagtail/admin/views/pages/edit.py [hunk 4]
```diff
@@ -32,6 +34,81 @@
...
+
+class HistorySidePanel(BaseSidePanel):
+    name = "history"
+    title = gettext_lazy("History")
+    template_name = "wagtailadmin/pages/editor_sidebar_panels/history.html"
+    order = 200
+
+    class Toggle(BaseSidePanel.Toggle):
+        aria_label = gettext_lazy("Toggle history")
+        icon_name = "history"
```
Comment (thibaudcolas) [inline wagtail/admin/views/pages/edit.py:86]: How are we going to reconcile this with the fact that for now we want History to be a link to the existing History page? Do we need to create a new `toggle.html` just for this, and move the existing link in `page_edit_header.html` to that template?

File: wagtail/admin/views/pages/edit.py [hunk 5]
```diff
@@ -32,6 +34,81 @@
...
+
+class StatusSidePanel(BaseSidePanel):
+    name = "status"
+    title = gettext_lazy("Status")
+    template_name = "wagtailadmin/pages/editor_sidebar_panels/status.html"
+    order = 100
+
+    class Toggle(BaseSidePanel.Toggle):
+        aria_label = gettext_lazy("Toggle status")
+        icon_name = "site"  # TODO Find the real icon
```
Comment (thibaudcolas) [inline wagtail/admin/views/pages/edit.py:75]: https://fontawesome.com/v5/icons/info-circle?s=solid

```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"/></svg>
```

File: wagtail/admin/views/pages/edit.py [hunk 6]
```diff
@@ -32,6 +34,81 @@
...
+
+class PreviewSidePanel(BaseSidePanel):
+    name = "preview"
+    title = gettext_lazy("Preview")
+    template_name = "wagtailadmin/pages/editor_sidebar_panels/preview.html"
+    order = 400
+
+    class Toggle(BaseSidePanel.Toggle):
+        aria_label = gettext_lazy("Toggle preview")
+        icon_name = "site"  # TODO Find the real icon
```
Comment (thibaudcolas) [inline wagtail/admin/views/pages/edit.py:109]: https://fontawesome.com/v5/icons/mobile-alt?s=solid

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><path d="M272 0H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h224c26.5 0 48-21.5 48-48V48c0-26.5-21.5-48-48-48zM160 480c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32zm112-108c0 6.6-5.4 12-12 12H60c-6.6 0-12-5.4-12-12V60c0-6.6 5.4-12 12-12h200c6.6 0 12 5.4 12 12v312z"/></svg>
```

File: wagtail/admin/views/pages/edit.py [hunk 7]
```diff
@@ -867,6 +944,12 @@ def get_context_data(self, **kwargs):
             instance=self.page, request=self.request, form=self.form
         )
         action_menu = PageActionMenu(self.request, view="edit", page=self.page)
+        side_panels = [
+            StatusSidePanel(self.page_for_status),
+            HistorySidePanel(self.page_for_status),
+            CommentsSidePanel(self.page_for_status),
+            PreviewSidePanel(self.page_for_status),
```
Comment (thibaudcolas) [inline wagtail/admin/views/pages/edit.py:951]: I imagine for now we’d need to remove this:

```suggestion
            # PreviewSidePanel(self.page_for_status),
```

File: wagtail/admin/views/pages/edit.py [hunk 8]
```diff
@@ -32,6 +34,81 @@
 )
 
 
+class BaseSidePanel(Component):
```
Comment (thibaudcolas) [inline wagtail/admin/views/pages/edit.py:37]: Eventually I think we will need all of those in CreateView as well, so should we move those definitions to a separate file already?

File: wagtail/admin/templates/wagtailadmin/pages/editor_sidebar_panels/toggle.html [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{% load wagtailadmin_tags %}
+
+<button class="w-bg-transparent w-py-3 w-px-3 w-inline-flex w-justify-center w-items-center w-outline-offset-inside w-text-grey-400 w-transition hover:w-transform hover:w-scale-110 hover:w-text-primary focus:w-text-primary"
+    aria-label="{{ toggle.aria_label }}"
+    data-tippy-content="{{ panel.title }}"
+    data-tippy-offset="[0, 6]"
+    data-tippy-placement="bottom"
+    data-side-panel-toggle="{{ panel.name }}"
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/pages/editor_sidebar_panels/toggle.html:8]:
```suggestion
    data-side-panel-toggle="{{ panel.name }}"
    aria-expanded="false"
```

File: wagtail/admin/templates/wagtailadmin/pages/editor_sidebar_panels/comments_toggle.html [hunk 10]
```diff
@@ -0,0 +1,18 @@
+{% load wagtailadmin_tags %}
+
+<button class="w-bg-transparent w-py-3 w-px-3 w-inline-flex w-justify-center w-items-center w-outline-offset-inside w-text-grey-400 w-transition hover:w-transform hover:w-scale-110 hover:w-text-primary focus:w-text-primary {% if comments_number %} w-mr-2.5 {% endif %}"
+    aria-label="{{ toggle.aria_label }}"
+    data-tippy-content="{{ panel.title }}"
+    data-tippy-offset="[0, 6]"
+    data-tippy-placement="bottom"
+    data-side-panel-toggle="{{ panel.name }}"
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/pages/editor_sidebar_panels/comments_toggle.html:8]:
```suggestion
    data-side-panel-toggle="{{ panel.name }}"
    aria-expanded="false"
```

File: wagtail/admin/templates/wagtailadmin/pages/editor_sidebar_panels/comments_toggle.html [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+>
+    {% icon name=toggle.icon_name class_name="w-w-4 w-h-4" %}
+
+    {% if comments_number %}
+        <div class="w-w-[10px] w-h-[10px] w-flex w-justify-center w-items-center w-p-[3px] w-font-bold w-absolute w-bg-teal-100 w-text-white w-text-[10px] w-border w-border-white w-rounded-full -w-top-[10px] -w-right-[12px]"
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/pages/editor_sidebar_panels/comments_toggle.html:13]:
```suggestion
        <div class="w-w-[10px] w-h-[10px] w-flex w-justify-center w-items-center w-p-[3px] w-font-bold w-absolute w-bg-teal-100 w-text-white w-text-[0.625rem] w-border w-border-white w-rounded-full -w-top-[10px] -w-right-[12px]"
```

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
Looking pretty good :) There are a few bugs to fix, and I’ve suggested a few refactorings & improvements, it’s not far off.

From my comments, the biggest points are:

- Whether this should also be present on the "create" view
- Compatibility with usage of `page-editor.js` without the side panels (right now there is a JS error in the console)
- How we make this work with the "History" link

In the interest of time I’d suggest applying all my line-by-line suggestions where possible, and then merge with just the above 3 points fixed.

File: wagtail/admin/views/pages/edit.py (hunk 8 above)
Comment (kaedroho) [reply wagtail/admin/views/pages/edit.py]: Done, thanks for pointing this out!

File: wagtail/admin/views/pages/edit.py (hunk 4 above)
Comment (kaedroho) [reply wagtail/admin/views/pages/edit.py]: Oh, I didn't know this. I think it should be a separate thing. I'll remove the history panel for now

File: wagtail/admin/views/pages/edit.py (hunk 6 above)
Comment (kaedroho) [reply wagtail/admin/views/pages/edit.py]: Thanks!

**Review by thibaudcolas — APPROVED** [review]
Thank you @kaedroho! Merging this now to unblock #8285. I’ll follow up on the focus restoration issue separately.

**Pull request merged**